rfcs/0049-flakes.md

```diff
@@ -0,0 +1,500 @@
+---
+feature: flakes
+start-date: 2019-07-09
+author: Eelco Dolstra
+co-authors: TBD
+shepherd-team: Domen Kožar, Alyssa Ross, Shea Levy, John Ericson
+shepherd-leader: Domen Kožar
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a mechanism to package Nix expressions into
+composable entities called "flakes". Flakes allow hermetic,
+reproducible evaluation of multi-repository Nix projects; impose a
+discoverable, standard structure on Nix projects; and replace previous
+mechanisms such as Nix channels and the Nix search path.
+
+# Motivation
+[motivation]: #motivation
+
+Flakes are motivated by a number of serious shortcomings in Nix:
+
+* While Nix pioneered reproducible builds, sadly, Nix expressions are
+  not nearly as reproducible as Nix builds. Nix expressions can access
+  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
+  environment variables, and Git repositories. This means for instance
+  that it is hard to ensure reproducible evaluation of NixOS or NixOps
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
+
+* There is no standard way to compose Nix projects. Typical ways are
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems. This RFC,
+however, only describes the format and semantics of flakes; it doesn't
+describe changes to the `nix` command to support flakes.
+
+# Detailed design
+[design]: #detailed-design
+
+## Flakes
+
+A flake is a directory that contains a file named `flake.nix` in the
```

Comment:
According to the current [implementation](https://github.com/NixOS/nix/blob/8c75621da6260d92459e105892751457e1659741/src/libexpr/flake/flakeref.cc#L126) a directory is only a flake if it is in a git repository:
```
error: path '/tmp/test-flake' is not a flake (because it does not reference a Git repository)
```
I can't see the wording here that requires using Git.

Is this a bug in the RFC or in the implementation?

Reply:
*ping* I'd be interested in this as well.

Reply:
This is no longer the case. You can also have a plain directory containing a `flake.nix` file.